tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+			break
+		}
+	}
+	if _, err := s.f.Seek(-bytesToOverwrite, 1); err != nil {
+		return err
+	}
+	for i := int64(0); i < bytesToOverwrite; i++ {
+		if _, err := s.f.Write([]byte{0x00}); err != nil {
+			return err
+		}
+	}
+	if err := s.f.Sync(); err != nil {
+		return err
+	}
+	_, err := s.f.Seek(-bytesToOverwrite, 1)
+	return err
+}
+
+// entry is used as a temporary object when reading data from segment files
+type entry struct {
+	key       []byte
+	data      []byte
+	timestamp int64
+}
+
+// cursor is a forward cursor for a given entry in the cache
+type cursor struct {
+	cache    [][]byte
+	position int
+}
+
+// Seek will point the cursor to the given time (or key)
+func (c *cursor) Seek(seek []byte) (key, value []byte) {
+	for i, p := range c.cache {
+		if bytes.Compare(seek, p[0:8]) >= 0 {
+			c.position = i
+			return p[0:8], p[8:]
+		}
+	}
+	return nil, nil
+}
+
+// Next moves the cursor to the next key/value. will return nil if at the end
+func (c *cursor) Next() (key, value []byte) {
+	pos := c.position + 1
+	if pos < len(c.cache) {
+		c.position = pos
+		v := c.cache[c.position]
+		return v[0:8], v[8:]
+	}
+	return nil, nil
+}
+
+func (w *Log) WritePoints(points []tsdb.Point) error {
+	partitionsToWrite := w.pointsToPartitions(points)
+
+	// get it to disk
+	if err := func() error {
+		w.mu.RLock()
+		defer w.mu.RUnlock()
+
+		for p, points := range partitionsToWrite {
+			if err := p.write(points); err != nil {
+				return err
+			}
+		}
+		return nil
+	}(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *Log) Flush() error {
+	return fmt.Errorf("explicit call to flush isn't implemented yet")
+}
+
+// pointsToPartitions returns a map that organizes the points into the partitions they should be mapped to
+func (w *Log) pointsToPartitions(points []tsdb.Point) map[*partition][]tsdb.Point {
+	m := make(map[*partition][]tsdb.Point)
+	for _, p := range points {
+		pp := w.walPartition(p.Key())
+		m[pp] = append(m[pp], p)
+	}
+	return m
+}
+
+// openPartitionFiles will open all partitions and read their segment files
+func (w *Log) openPartitionFiles() error {
+	results := make(chan error, len(w.partitions))
+	for _, p := range w.partitions {
+
+		go func(p *partition) {
+			fileNames, err := p.segmentFileNames()
+			if err != nil {
+				results <- err
+				return
+			}
+			for _, n := range fileNames {
+				entries, err := p.readFile(n)
+				if err != nil {
+					results <- err
+					return
+				}
+				for _, e := range entries {
+					p.addToCache(e.key, e.data, e.timestamp)
+				}
+			}
+			results <- nil
+		}(p)
+	}
+
+	for i := 0; i < len(w.partitions); i++ {
+		err := <-results
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Close will finish any flush that is currently in process and close file handles
+func (w *Log) Close() error {
+	// stop the autoflushing process so it doesn't try to kick another one off
+	if w.closing != nil {
+		close(w.closing)
+		w.closing = nil
+	}
```

Comment:
I was doing it that way before, but it was creating a race condition with the `autoFlusher`. Basically this would get the mutex, then the autoFlusher would start before the channel was close and the autoFlusher was waiting on a mutex to see if it should flush, causing a deadlock.

Maybe use a separate mutex just for closing?